sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala

```diff
@@ -41,18 +46,107 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
   override def output: Seq[Attribute] = Nil
 }
 
+case class AppendDataExec(
+    table: SupportsBatchWrite,
+    writeOptions: DataSourceOptions,
+    query: SparkPlan) extends V2TableWriteExec with BatchWriteHelper {
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    val batchWrite = newWriteBuilder() match {
+      case builder: SupportsSaveMode => // TODO: Remove this
+        builder.mode(SaveMode.Append).buildForBatch()
+
+      case builder =>
+        builder.buildForBatch()
+    }
+    doWrite(batchWrite)
+  }
+}
+
+case class OverwriteByExpressionExec(
+    table: SupportsBatchWrite,
+    filters: Array[Filter],
+    writeOptions: DataSourceOptions,
+    query: SparkPlan) extends V2TableWriteExec with BatchWriteHelper {
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    val batchWrite = newWriteBuilder() match {
+      case builder: SupportsOverwrite =>
+        builder.overwrite(filters).buildForBatch()
+
+      case builder: SupportsSaveMode => // TODO: Remove this
+        builder.mode(SaveMode.Overwrite).buildForBatch()
+
+      case _ =>
+        throw new SparkException(s"Table does not support dynamic partition overwrite: $table")
```

Comment:
`AnalysisException`?

Reply:
Other physical plans throw SparkException. I thought AnalysisException was just for query analysis?

Reply:
ditto. (https://github.com/apache/spark/pull/23606/files#r251616435)

Reply:
This is definitely a case where it is too late to throw an analysis exception. This will be checked by a new analysis rule when capabilities are added.

Reply:
Here we are executing a plan, so `AnalysisException` should not be used. I'm still worried about `SparkException` because we are not launching a Spark job here.

Shall we throw `IllegalStateException`? Eventually we will make this never happen, as this kind of check should be done as early as possible.